<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.

Dataset information: https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks)
Download it from here: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

In [71]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [72]:
import warnings
warnings.filterwarnings('ignore')

In [73]:
df = pd.read_csv("./sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
199  0.0116  0.0744  0.0367  0.0225  0.0076  0.0545  0.1110  0.1069  0.1708   
174  0.0191  0.0173  0.0291  0.0301  0.0463  0.0690  0.0576  0.1103  0.2423   
31   0.0084  0.0153  0.0291  0.0432  0.0951  0.0752  0.0414  0.0259  0.0692   
129  0.1371  0.1226  0.1385  0.1484  0.1776  0.1428  0.1773  0.2161  0.1630   
126  0.0715  0.0849  0.0587  0.0218  0.0862  0.1801  0.1916  0.1896  0.2960   

         9   ...      51      52      53      54      55      56      57  \
199  0.2271  ...  0.0141  0.0103  0.0100  0.0034  0.0026  0.0037  0.0044   
174  0.3134  ...  0.0040  0.0136  0.0137  0.0172  0.0132  0.0110  0.0122   
31   0.1753  ...  0.0236  0.0114  0.0136  0.0117  0.0060  0.0058  0.0031   
129  0.2067  ...  0.0118  0.0129  0.0344  0.0065  0.0067  0.0022  0.0079   
126  0.4186  ...  0.0153  0.0121  0.0096  0.0196  0.0042  0.0066  0.0099   

         58      59  60  
199  0.0057  0.0035   M  
174  0.0114  0.0068   M  
31   0.0072  0.0045   R  
129  0.0146  0.0051   M  
126  0.0083  0.0124   M  

[5 rows x 61 columns]

In [74]:
df.shape

(208, 61)

In [75]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [76]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [77]:
df[60].value_counts() # label is not skewed

M    111
R     97
Name: 60, dtype: int64

In [78]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [79]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
59   1
101  0
184  0
141  0
191  0

In [80]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [81]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [82]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [83]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

# Using Deep Learning Model

## Model without Dropout Layer

In [84]:
import tensorflow as tf
from tensorflow import keras

In [85]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6895 - accuracy: 0.5321
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6669 - accuracy: 0.5833
Epoch 3/100
20/20 [==============================] - 0s 1ms/step - loss: 0.6512 - accuracy: 0.6474
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6189 - accuracy: 0.6923
Epoch 5/100
20/20 [==============================] - 0s 1ms/step - loss: 0.5780 - accuracy: 0.7564
Epoch 6/100
20/20 [==============================] - 0s 1ms/step - loss: 0.5420 - accuracy: 0.8141
Epoch 7/100
20/20 [==============================] - 0s 1ms/step - loss: 0.5194 - accuracy: 0.7756
Epoch 8/100
20/20 [==============================] - 0s 1ms/step - loss: 0.4909 - accuracy: 0.7885
Epoch 9/100
20/20 [==============================] - 0s 1ms/step - loss: 0.4450 - accuracy: 0.8013
Epoch 10/100
20/20 [==============================] - 0s 1ms/step - loss: 0.4145 - accuracy: 0.8462
Epoch 11/

20/20 [==============================] - 0s 1ms/step - loss: 0.0071 - accuracy: 1.0000
Epoch 84/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0067 - accuracy: 1.0000
Epoch 85/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0064 - accuracy: 1.0000
Epoch 86/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0061 - accuracy: 1.0000
Epoch 87/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0063 - accuracy: 1.0000
Epoch 88/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0057 - accuracy: 1.0000
Epoch 89/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0057 - accuracy: 1.0000
Epoch 90/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0053 - accuracy: 1.0000
Epoch 91/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0055 - accuracy: 1.0000
Epoch 92/100
20/20 [==============================] - 0s 1ms/step - loss: 0.0049 - accuracy: 1.0000
Epoch 93/100


In [86]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 1ms/step - loss: 0.8629 - accuracy: 0.7692


[0.8629363775253296, 0.7692307829856873]

Training Accuracy >>> Test Accuracy

In [87]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[7.1083555e-08 3.7778303e-01 9.9051148e-01 2.6755422e-05 9.9999976e-01
 9.9999321e-01 6.4399661e-03 9.9999988e-01 9.1076899e-06 9.9999988e-01]
[0. 0. 1. 0. 1. 1. 0. 1. 0. 1.]


In [88]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [89]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.71      0.93      0.81        27
           1       0.88      0.60      0.71        25

    accuracy                           0.77        52
   macro avg       0.80      0.76      0.76        52
weighted avg       0.80      0.77      0.76        52



### Model with Dropout Layer

In [90]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7653 - accuracy: 0.4744
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6923 - accuracy: 0.5256
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7149 - accuracy: 0.4872
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7439 - accuracy: 0.4423
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6974 - accuracy: 0.5064
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7053 - accuracy: 0.5000
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7035 - accuracy: 0.4808
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6994 - accuracy: 0.4679
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6984 - accuracy: 0.5385
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6833 - accuracy: 0.5000
Epoch 11/

20/20 [==============================] - 0s 2ms/step - loss: 0.3426 - accuracy: 0.8590
Epoch 84/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3245 - accuracy: 0.9038
Epoch 85/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3396 - accuracy: 0.8526
Epoch 86/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3587 - accuracy: 0.8590
Epoch 87/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3705 - accuracy: 0.8205
Epoch 88/100
20/20 [==============================] - 0s 2ms/step - loss: 0.2835 - accuracy: 0.8654
Epoch 89/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3712 - accuracy: 0.8013
Epoch 90/100
20/20 [==============================] - 0s 1ms/step - loss: 0.3640 - accuracy: 0.8333
Epoch 91/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3402 - accuracy: 0.8974
Epoch 92/100
20/20 [==============================] - 0s 1ms/step - loss: 0.3445 - accuracy: 0.8654
Epoch 93/100


In [91]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 1ms/step - loss: 0.3962 - accuracy: 0.8077


[0.39620405435562134, 0.807692289352417]

Training Accuracy is still good but Test Accuracy Improved

In [92]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[0.00204679 0.855612   0.87832576 0.02817929 0.99947876 0.92398506
 0.24040593 0.99971884 0.04001041 0.99982685]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [93]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.76      0.93      0.83        27
           1       0.89      0.68      0.77        25

    accuracy                           0.81        52
   macro avg       0.83      0.80      0.80        52
weighted avg       0.82      0.81      0.80        52



**You can see that by using dropout layer test accuracy increased from 0.77 to 0.81**